op-supervisor/supervisor/backend/db/db.go

```diff
@@ -0,0 +1,494 @@
+	}
+	// Read all entries until the end of the file
+	for {
+		_, _, _, err := i.NextLog()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return fmt.Errorf("failed to init from existing entries: %w", err)
+		}
+	}
+	db.lastEntryContext = logContext{
+		blockNum: i.blockNum,
+		logIdx:   i.logIdx,
+	}
+	return nil
+}
+
+func (db *DB) updateEntryCountMetric() {
+	db.m.RecordEntryCount(db.lastEntryIdx + 1)
+}
+
+// ClosestBlockInfo returns the block number and hash of the highest recorded block at or before blockNum.
+// Since block data is only recorded in search checkpoints, this may return an earlier block even if log data is
+// recorded for the requested block.
+func (db *DB) ClosestBlockInfo(blockNum uint64) (uint64, TruncatedHash, error) {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	checkpointIdx, err := db.searchCheckpoint(blockNum, math.MaxUint32)
+	if err != nil {
+		return 0, TruncatedHash{}, fmt.Errorf("no checkpoint at or before block %v found: %w", blockNum, err)
+	}
+	checkpoint, err := db.readSearchCheckpoint(checkpointIdx)
+	if err != nil {
+		return 0, TruncatedHash{}, fmt.Errorf("failed to reach checkpoint: %w", err)
+	}
+	canonicalHash, err := db.readCanonicalHash(checkpointIdx + 1)
+	if err != nil {
+		return 0, TruncatedHash{}, fmt.Errorf("failed to read canonical hash: %w", err)
+	}
+	return checkpoint.blockNum, canonicalHash, nil
+}
+
+// Contains return true iff the specified logHash is recorded in the specified blockNum and logIdx.
+// logIdx is the index of the log in the array of all logs the block.
+func (db *DB) Contains(blockNum uint64, logIdx uint32, logHash common.Hash) (bool, error) {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	db.log.Trace("Checking for log", "blockNum", blockNum, "logIdx", logIdx, "hash", truncateHash(logHash))
+	entryIdx, err := db.searchCheckpoint(blockNum, logIdx)
+	if errors.Is(err, io.EOF) {
+		// Did not find a checkpoint to start reading from so the log cannot be present.
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	i, err := db.newIterator(entryIdx)
+	if err != nil {
+		return false, fmt.Errorf("failed to create iterator: %w", err)
+	}
+	db.log.Trace("Starting search", "entry", entryIdx, "blockNum", i.blockNum, "logIdx", i.logIdx)
+	defer func() {
+		db.m.RecordSearchEntriesRead(i.entriesRead)
+	}()
+	for {
+		evtBlockNum, evtLogIdx, evtHash, err := i.NextLog()
+		if errors.Is(err, io.EOF) {
+			// Reached end of log without finding the event
+			return false, nil
+		} else if err != nil {
+			return false, fmt.Errorf("failed to read next log: %w", err)
+		}
+		if evtBlockNum == blockNum && evtLogIdx == logIdx {
+			db.log.Trace("Found initiatingEvent", "blockNum", evtBlockNum, "logIdx", evtLogIdx, "hash", evtHash)
+			// Found the requested block and log index, check if the hash matches
+			return slices.Equal(evtHash, truncateHash(logHash)), nil
```

Comment:
It's truncating the hash here again, I think the above suggestion of utilizing the truncated-hash type, and providing that as function argument, makes it cleaner.